### PR TITLE
added json completion

### PR DESCRIPTION
**Pull request by imolorhe**

### CODE REVIEW

**Comment by acao** [comment]
wow! this is all it took then?? this is so exciting to see LSP libraries being used without requiring an lsp compatibility layer

**Review by acao — APPROVED** [review]
This is awesome! I will add a deploy and tests, and linting

File: src/json-completion.ts [hunk 1]
```diff
@@ -0,0 +1,815 @@
...
+  add(completion: Completion) {
+    if (this.reservedKeys.has(completion.label)) {
+      return;
+    }
+    this.completions.set(completion.label, completion);
+  }
+}
+
+export class JSONCompletion {
+  constructor(private schema: JSONSchema7) {}
```
Comment (acao) [inline src/json-completion.ts:51]: this is a good starting point!

File: src/json-completion.ts [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+
+  // apply is the quoted key which will be applied.
+  // Normally the label needs to match the token
+  // prefix i.e. if the token begins with `"to`, then the
+  // label needs to have the quotes as well for it to match.
+  // However we are manually filtering the results so we can
+  // just use the unquoted key as the label, which is nicer
+  // and gives us more control.
+  // If no property value is present, then we add the colon as well.
+  // TODO: Use snippetCompletion to handle insert value + position cursor e.g. "key": "#{}"
```
Comment (acao) [inline src/json-completion.ts:248]: ahhh, i see, yes in monaco this is $0 or $1, this makes sense

File: src/json-completion.ts [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+          case 'null':
+            value = '#{null}';
+            break;
+          default:
+            return resultText;
+        }
+      }
+    }
+    if (!value || nValueProposals > 1) {
+      value = '$1';
```
Comment (acao) [inline src/json-completion.ts:332]: i think we just need to change this to `'#{}'` yes?

File: src/json-completion.ts [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+  }
+
+  private getInsertTextForGuessedValue(
+    value: any,
+    separatorAfter = ''
+  ): string {
+    switch (typeof value) {
+      case 'object':
+        if (value === null) {
+          return '${null}' + separatorAfter;
```
Comment (acao) [inline src/json-completion.ts:348]: i think these other `${}` insertTexts are still from monaco, yes?

Comment (acao) [reply src/json-completion.ts:348]: ok i think i understand a bit better after reading this

https://discuss.codemirror.net/t/autocomplete-cursor-position-in-apply-function/4088/3

**Comment by acao** [comment]
love it! awesome work

cm6-completion

File: src/json-completion.ts (hunk 3 above)
Comment (imolorhe) [reply src/json-completion.ts:332]: Yup!

File: src/json-completion.ts (hunk 4 above)
Comment (imolorhe) [reply src/json-completion.ts:348]: Yeah # or $ are interchangeable in codemirror snippet syntax (at least according to the docs)

**Pull request merged**